### DISCUSSION TITLE
JDim 0.11.0-20240113

### DISCUSSION THREAD

**Discussion by ma8ma** [JDimproved/JDim#1323] [General]
## 主な変更点

詳細は [マージされたPull requests][prs] または [マニュアル][m] を参照してください。  
リリース計画のissue: [#1269]

[prs]: https://github.com/JDimproved/JDim/pulls?utf8=%E2%9C%93&q=merged%3A2023-07-23..2024-01-13+is%3Apr+is%3Amerged+base%3Amaster+
[m]: https://jdimproved.github.io/JDim/
[#1269]: https://github.com/JDimproved/JDim/issues/1269

### :point_right: ユーザーインターフェース
- ([#1215]) ショートカットキーとマウスジェスチャの詳細設定に "右(左)の更新可能なタブへ移動" を追加しました。
- ([#1238]) 発言(同一ID)数をカウントして表示する設定のときはレスが何番目の投稿なのか分かるように (何番目の投稿/発言数) の形式に変更しました。
  **注意**:
  1000レスを超える長大なスレッドや一度に大量のスレッドを開いたり、古いPCを利用している場合は実行速度に影響があるかもしれません。
- ([#1240]) スレッドの投稿者IDについて発言数の表示を最大5桁にしました。5桁より大きいときは99999と表示します。
- ([#1246]) まちBBSでofflaw.cgiを使わないモードでスレッドを読み込んで
  通常リンク(青)をクリックしたときリンク先の確認ページではなく直接ページを表示するように修正しました。
- ([#1252]) キャッシュ内ログ検索の結果はスレが立った時刻が早い順に並べ替えて表示するように変更しました。
- ([#1257]) ディスプレイ解像度が小さい環境でウインドウを開いたときデスクトップ画面からはみ出さないようにサイズを調整しました。
- ([#1259]) about:config に「スレビューのテキストを描画する方法 ( 0: PangoGlyphString 1: PangoLayout )」を追加しました。
  また、ビルドオプション(`-Dpangolayout=enabled`)はデフォルト設定を指定するように変更しました。
- ([#1279]) 板のプロパティにある「あぼ〜ん設定(スレビュー)」に「連続投稿したIDをスレのNG IDに追加する」設定を追加しました。
  設定可能な値は 0(未設定) から 1000 までの範囲で初期設定は 0 です。
  **注意**:
  連続投稿したIDをあぼーんする設定が優先されるためスレのプロパティにあるNG IDからIDを削除しても再び登録されることがあります。
  このオプションは実験的なサポートのため変更または廃止の可能性があります。🧪
- ([#1307]) 板一覧に外部BBSMENUを登録する機能を追加しました。
  詳細は[オンラインマニュアル][etcbbsmenu]を参照してください。
  この機能は実験的なサポートのため変更または廃止の可能性があります。🧪

[#1215]: https://github.com/JDimproved/JDim/pull/1215
[#1238]: https://github.com/JDimproved/JDim/pull/1238
[#1240]: https://github.com/JDimproved/JDim/pull/1240
[#1246]: https://github.com/JDimproved/JDim/pull/1246
[#1252]: https://github.com/JDimproved/JDim/pull/1252
[#1257]: https://github.com/JDimproved/JDim/pull/1257
[#1259]: https://github.com/JDimproved/JDim/pull/1259
[#1279]: https://github.com/JDimproved/JDim/pull/1279
[#1307]: https://github.com/JDimproved/JDim/pull/1307
[etcbbsmenu]: https://jdimproved.github.io/JDim/external/#register_bbsmenu

### 🔠 スレ一覧とスレビューのテキストエンコーディング

- ([#1266]) about:config の「(安全でない) スレ一覧とスレビューのプロパティにあるエンコーディング設定を有効にする」
  が"はい"のときはHTTP通信のエンコーディング情報は無視して設定された値を優先するように変更しました。
- ([#1267]) 板のプロパティに「UTF-8で書き込む」実験的なオプションを追加しました。
  掲示板がUTF-8の書き込みに対応してるか確認して使用してください。
  このオプションは実験的なサポートのため変更または廃止の可能性があります。🧪
- ([#1272]) 板のプロパティにエンコーディングを判定する方法を選択するオプションを実装しました。(下記参照)
- ([#1277]) 板のプロパティにあるエンコーディングを判定する方法に「テキストからエンコーディングを推測する」オプションを追加しました。
  エンコーディングを推測した結果が正しいとは限らないため利用の際は注意してください。

#### 板のプロパティに追加されたエンコーディングを判定する方法について
この設定は about:config の「(安全でない) スレ一覧とスレビューのプロパティにあるエンコーディング設定を有効にする」が"はい"のときは利用できません。
このオプションは実験的なサポートのため変更または廃止の可能性があります。🧪

- ([#1272]) デフォルト設定を使う (初期設定)
- (実装済) HTTPヘッダーのエンコーディング情報を使う
- ([#1277]) テキストからエンコーディングを推測する

**互換性の注意**:
修正前の動作は「HTTPヘッダーのエンコーディング情報を使う」ですが、修正後は「デフォルト設定を使う」が初期設定になり動作が変更されます。

[#1266]: https://github.com/JDimproved/JDim/pull/1266
[#1267]: https://github.com/JDimproved/JDim/pull/1267
[#1272]: https://github.com/JDimproved/JDim/pull/1272
[#1277]: https://github.com/JDimproved/JDim/pull/1277

### :globe_with_meridians: ネットワーク通信
- ([#1216]) プロキシ設定に5ch過去ログの読み込みに失敗したとき2ch読み込み用プロキシを使うオプションを追加しました。
  このオプションは実験的なサポートのため変更または廃止の可能性があります。🧪
- ([#1230]) 上記 #1216 で追加したオプションの動作を修正して過去ログサーバ(サーバ移転前のURL)のプロキシ接続に対応しました。
- ([#1280]) ソケットの読み込み、書き込みが準備完了するまで待機する処理にpoll(2)を使うように変更しました。

[#1216]: https://github.com/JDimproved/JDim/pull/1216
[#1230]: https://github.com/JDimproved/JDim/pull/1230
[#1280]: https://github.com/JDimproved/JDim/pull/1280

[20230711]: https://agree.5ch.net/test/read.cgi/operate/9240230711/3

### :package: Snapパッケージ
no changes

### :sparkles: 他に変更・追加された機能
- ([#1239]) スレッドの投稿者IDの発言数を数える処理の効率を改善しました。ただし、コンパイラーが最適化を行っている場合は効果は微小です。
- ([#1294]) 5ch.netの板と同名の板を持つおーぷん２ちゃんねる(open2ch.net)、Next2ch(next2ch.net)、2ちゃんねる(2ch.sc)は板移転の対象にならないように修正しました。

[#1239]: https://github.com/JDimproved/JDim/pull/1239
[#1294]: https://github.com/JDimproved/JDim/pull/1294

### :lady_beetle: バグ修正
- ([#1213], [#1214]) 空白行を挟んだ書き込みがあると、自分の書き込みと認識されない不具合を修正しました。 Thanks @mtasaka for report and PR.
- ([#1221]) 先頭に空行があるレスを5chのスレに書き込んだとき書き込みマークが付かない不具合を修正しました。 Thanks @mtasaka for PR.
- ([#1222]) 2chスレ過去ログのスレにあるレスアンカーの上にマウスオーバーしてもポップアップが出ない不具合を修正しました。
- ([#1226]) 板のプロパティにあるローカルルールのリンクが表示されない不具合を修正しました。
- ([#1237]) まちBBSのスレッドにある通常リンク(青色)をクリックしたときブラウザで正常に開けない不具合を修正しました。
- ([#1245]) まちBBSの板のうち地方名サブドメインが付いてないURLを登録して開くと読み込みに失敗する不具合を修正しました。
- ([#1247]) 5ch.netの板のうちサブドメインが付いてないURLのリンクをクリックするとwebブラウザで開かれる問題を修正しました。
- ([#1251]) キャッシュ内ログ検索の実行を中止できない不具合を修正しました。
- ([#1256]) 関数を呼び出した後にerrnoの値をチェックする箇所では関数を呼び出す前にerrnoを0にリセットするように修正しました。
- ([#1276]) スレ一覧のタブをクリックしたときにクラッシュすることがある問題を修正しました。
- ([#1281]) 板やスレの読み込みを中止しても通信がキャンセルされないことがある不具合を修正しました。
- ([#1303]) migemoとUnity buildを組み合わせてビルドするとコンパイルエラーになる場合があったため修正しました。
- ([#1320]) 書き込みログをファイルに保存するときスレタイが通常のテキストで表示されるように修正しました。
  **注意**: これまでに保存されたログは文字参照のまま変わらず表示されるため
  過去のログを直すにはログファイル(log/postlog)を手動で編集する必要があります。

[#1213]: https://github.com/JDimproved/JDim/pull/1213
[#1214]: https://github.com/JDimproved/JDim/pull/1214
[#1221]: https://github.com/JDimproved/JDim/pull/1221
[#1222]: https://github.com/JDimproved/JDim/pull/1222
[#1226]: https://github.com/JDimproved/JDim/pull/1226
[#1237]: https://github.com/JDimproved/JDim/pull/1237
[#1245]: https://github.com/JDimproved/JDim/pull/1245
[#1247]: https://github.com/JDimproved/JDim/pull/1247
[#1251]: https://github.com/JDimproved/JDim/pull/1251
[#1256]: https://github.com/JDimproved/JDim/pull/1256
[#1276]: https://github.com/JDimproved/JDim/pull/1276
[#1281]: https://github.com/JDimproved/JDim/pull/1281
[#1303]: https://github.com/JDimproved/JDim/pull/1303
[#1320]: https://github.com/JDimproved/JDim/pull/1320

### :green_book: ドキュメント
- ([#1299]) READMEとマニュアルにあるライセンスの説明を更新してGPL-2.0-or-laterを導入することを記載しました。
- ([#1299]) [RFC 0013][rfc13]が承認されたためコントリビュートのガイドにファイルのライセンスについて説明を追加しました。
- ([#1299]) リポジトリ用のプルリクエストテンプレートを追加しました。

[#1299]: https://github.com/JDimproved/JDim/pull/1299

### :wrench: ソフトウェアの構築（ビルド）
- ([#1217]) OpenBSDでビルドが失敗する問題を修正しました。

[#1217]: https://github.com/JDimproved/JDim/pull/1217

### :recycle: 継続的インテグレーション(CI)
- ([#1303], [#1306]) 定期的に実行する Weekly CI にオプションの依存関係とUnity buildを有効にしてビルドするjobを追加しました。

[#1303]: https://github.com/JDimproved/JDim/pull/1303
[#1306]: https://github.com/JDimproved/JDim/pull/1306

[buster-bpo-meson]: https://packages.debian.org/buster-backports/meson
[buster-bpo-gtest]: https://packages.debian.org/buster-backports/libgtest-dev
[unitybuild]: https://mesonbuild.com/Unity-builds.html

### :ghost: 廃止された機能
- ([#1209]) ビルドツールAutotools(./configure)のサポートを廃止して関連のスクリプトファイルを削除しました。([RFC 0012][rfc12])

[#1209]: https://github.com/JDimproved/JDim/pull/1209
[rfc12]: https://github.com/JDimproved/rfcs/blob/master/docs/0012-end-of-autotools-support.md

### :page_with_curl: ライセンス
- ([RFC 0013][rfc13]) **2023-12-09 以降に取り込まれた修正(パッチやコミット)は GPL-2.0-or-later でライセンスされます。**
  また、寛容なライセンスが使われているファイルの修正にはそのライセンスが適用されます。
- ([#1298]) JDimのプログラムに影響するソースコードのうちファイルの修正に関わった貢献者全員から承諾を得たものは
  ライセンスをGPL-2.0-or-laterに変更しました。
- ([#1298]) JDimのプログラムに影響しない設定ファイルなどでファイルの修正に関わった貢献者全員から承諾を得られたものは
  ライセンスをFSFによって作成された単純な寛容ライセンス(FSFAP)に変更しました。
- ([#1298]) JDimのプログラムに影響しないマニュアルのうちファイルの修正に関わった貢献者全員から承諾を得たものは
  ライセンスをFSFAPとGPL-2.0-or-laterのデュアルライセンスに変更しました。

[#1298]: https://github.com/JDimproved/JDim/pull/1298
[rfc13]: https://github.com/JDimproved/rfcs/blob/87d3c6e1d9/docs/0013-introduce-license-gpl-2.0-or-later.md

### :soon: サポート
- ([#1210], [#1231]) 以下のプラットフォームのサポートは0.11.0リリースをもって終了します。
  - Ubuntu 20.04(2020年リリース)より古いディストロ ([RFC 0006][rfc6])
  - コンパイラのバージョンが gcc-9 未満または clang-10 未満のディストロ

[#1210]: https://github.com/JDimproved/JDim/pull/1210
[#1231]: https://github.com/JDimproved/JDim/pull/1231
[rfc6]: https://github.com/JDimproved/rfcs/blob/d6469b24bc/docs/0006-platform-support.md

不具合のご報告やパッチのご提供をいただきまして重ねてお礼申し上げます。

<hr /><em>This discussion was created from the release <a href='https://github.com/JDimproved/JDim/releases/tag/JDim-v0.11.0'>JDim 0.11.0-20240113</a>.</em>

**Comment by ma8ma** [comment]
GitHubのリポジトリにあるreleaseページにはリポジトリから生成するtarballがありますが、
メンテナーが自らtarballを作成し電子署名(PGP)をつけてreleaseページで配布することがあります。
JDimでも v0.7.0 のリリースから私 @ma8ma (メンテナー)が署名したtarballをreleaseページにアップロードしています。

ところが、データ圧縮ツール [xz](https://github.com/tukaani-project/xz) (2024-03-30現在 _disabled_) の
releaseページで配布されていた署名付きtarballにバックドアが仕掛けられていたようです。

- https://nvd.nist.gov/vuln/detail/CVE-2024-3094
- https://thehackernews.com/2024/03/urgent-secret-backdoor-found-in-xz.html
- https://lwn.net/Articles/967180/
- https://gist.github.com/thesamesam/223949d5a074ebc3dce9ee78baad9e27

~~今回の件を受けて私が作成して署名したtarballよりリポジトリから生成するほうが信頼性があると考えています。
そこで次のリリースから署名付きtarballの配布を止めたいと思います。~~ 何か意見や要望などございましたらコメントいただけると幸いです。

**Reply by mtasaka** [reply]
まあでも例えば Fedora で配布しているバイナリrpmは、Fedora Projectで署名したものを配布していて、ユーザーはきちんと署名されているかを確認してinstall / update するようになっています。むしろ署名されてないrpmは出所不明のものとしてdefaultではinstallが拒否されますし。
普通は作者なり配布する人が署名したものの方が信頼出来ると考えるのが自然だと思います。まあxzの件は（恣意的だとすればですが）2年くらいかけて信頼を築いた上で裏切ったわけですし、githubで公開しているコードにもやろうと思えば悪意あるコードを埋め込めるのは変わらないので、結局は作者を信頼できるかどうかというのは変わらない気がします。

**Reply by ma8ma** [reply]
今回の件は引き継いだ私にも当てはまるところがあると思い、このままで信頼に足るのか痛いところを突かれた感じです 😨
今の状態だと署名付きtarballの中身を各自チェックしてくださいと伝えるしかないですがあったほうがいいですかね…

**Reply by mtasaka** [reply]
やっぱり署名付きtarballというのはあった方がいいとは思います。普通は書名は文字通り「作者が署名しましたよ」を示すものであるので。
今回のxzの件みたいに（確定ではないですが）意図的に不正コードを紛れ込ませた場合は、結局は誰か別の人が発見するしかないですし（まあ今回のxzの件は衝撃ですが...）

**Reply by ma8ma** [reply]
そうですね、メンテナー本人が保証することを示す証として署名付きtarballは継続します。
ソースコードのチェックは @mtasaka さんやjdimに関心のある皆さんと協力しながらやっていければと思います。
驚いて迷ってしまいましたが助かりました。ご意見いただきありがとうございます。🙇
